### ISSUE TITLE
workflow template ref does not work in other namespace

### ISSUE THREAD

**Issue by dove-young** [argoproj/argo#3759]
## Summary 

Reference workflow template from namespace other than `argo` dose not work. 

Submit workflow which reference template in `argo` namespace works .

## Diagnostics

What version of Argo Workflows are you running?

```
[root@sre4-inf examples]# argo version
argo: v2.9.5
  BuildDate: 2020-08-06T22:11:59Z
  GitCommit: 5759a0e198d333fa8c3e0aeee433d93808c0dc72
  GitTreeState: clean
  GitTag: v2.9.5
  GoVersion: go1.13
  Compiler: gc
  Platform: linux/amd64
[root@sre4-inf examples]#
```

### Reproduce 

- Install argo from https://github.com/argoproj/argo/blob/master/manifests/install.yaml
- Create a new namespace `kubectl create ns my-namespace`
- Create workflow template in `my-namespace` from this https://github.com/argoproj/argo/blob/master/examples/workflow-template/templates.yaml
- Create workflow  in `my-namespace` from this https://github.com/argoproj/argo/blob/master/examples/workflow-template/workflow-template-ref-with-entrypoint-arg-passing.yaml

```
Paste the logs from the workflow controller:
kubectl logs -n argo $(kubectl get pods -l app=workflow-controller -n argo -o name) | grep ${workflow}
```

```
[root@sre4-inf workflow-template]# argo submit  workflow-template-ref-with-entrypoint-arg-passing.yaml -n  my-namespace
Name:                workflow-template-hello-world-q25zd
Namespace:            my-namespace
ServiceAccount:      default
Status:              Pending
Created:             Wed Aug 12 19:04:12 -0700 (1 minute from now)
Parameters:
  message:           hello world
[root@sre4-inf workflow-template]#

[root@sre4-inf workflow-template]# argo get workflow-template-hello-world-q25zd
Name:                workflow-template-hello-world-q25zd
Namespace:            my-namespace
ServiceAccount:      default
Status:              Error
Message:             workflowtemplates.argoproj.io "workflow-template-whalesay-template" not found
Conditions:
 Completed           True
Created:             Wed Aug 12 19:04:12 -0700 (2 minutes ago)
Started:             Wed Aug 12 19:04:12 -0700 (2 minutes ago)
Finished:            Wed Aug 12 19:04:12 -0700 (2 minutes ago)
Duration:            0 seconds
Parameters:
  message:           hello world

[root@sre4-inf workflow-template]# argo template list
NAME
workflow-template-inner-dag
workflow-template-inner-steps
workflow-template-random-fail-template
workflow-template-submittable
workflow-template-whalesay-template
[root@sre4-inf workflow-template]#
```
---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by dove-young** [comment]
Submit ordinary workflow to other namespace works. 

```
[root@sre4-inf examples]# argo submit --watch  hello-world.yaml
Name:                hello-world-sck54
Namespace:           my-namespace
ServiceAccount:      default
Status:              Running
Created:             Wed Aug 12 19:13:26 -0700 (27 seconds from now)
Started:             Wed Aug 12 19:13:26 -0700 (27 seconds from now)
Duration:            27 seconds

STEP                  TEMPLATE  PODNAME            DURATION  MESSAGE
 ◷ hello-world-sck54  whalesay  hello-world-sck54  27s       ContainerCreating
^C
[root@sre4-inf examples]#
```

**Comment by sarabala1979** [comment]
`WorkflowTemplate` is a namespaced scope.   `WorkflowTemplate` and `Workflow` have to be in same namespace.  if you want to reference the workflowtemplate across the namespaces in the cluster, you need to use `ClusterWorkflowTemplate which is a cluster scope.

**Comment by dove-young** [comment]
They are in the same namespace

```
[root@sre4-inf workflow-template]# argo template list -n my-namespace
NAME
workflow-template-inner-dag
workflow-template-inner-steps
workflow-template-random-fail-template
workflow-template-submittable
workflow-template-whalesay-template
[root@sre4-inf workflow-template]#
[root@sre4-inf workflow-template]#
[root@sre4-inf workflow-template]# argo list -n my-namespace
NAME                                  STATUS   AGE   DURATION   PRIORITY
workflow-template-hello-world-5nsd2   Error    2h    0s         0
[root@sre4-inf workflow-template]# argo get workflow-template-hello-world-5nsd2 -n my-namespace
Name:                workflow-template-hello-world-5nsd2
Namespace:           my-namespace
ServiceAccount:      default
Status:              Error
Message:             workflowtemplates.argoproj.io "workflow-template-submittable" not found
Conditions:
 Completed           True
Created:             Wed Aug 12 19:48:44 -0700 (2 hours ago)
Started:             Wed Aug 12 19:48:44 -0700 (2 hours ago)
Finished:            Wed Aug 12 19:48:44 -0700 (2 hours ago)
Duration:            0 seconds
[root@sre4-inf workflow-template]#
```

**Comment by sarabala1979** [comment]
Can you share ‘workflow-controller’ log? You can get the using kubectl logs <workflow-controller pod>

**Comment by dove-young** [comment]
I enabled `--loglevel debug` and `gloglevel 9` , here is the result

```
time="2020-08-13T05:47:35Z" level=info msg="Processing workflow" namespace=stackstorm workflow=workflow-template-hello-world-lf6rp
time="2020-08-13T05:47:35Z" level=error msg="Unable to get Workflow Template Reference for workflow" error="workflowtemplates.argoproj.io \"workflow-template-submittable\" not found" namespace=stacksto
rm workflow=workflow-template-hello-world-lf6rp
time="2020-08-13T05:47:35Z" level=info msg="Updated phase  -> Error" namespace=stackstorm workflow=workflow-template-hello-world-lf6rp
time="2020-08-13T05:47:35Z" level=info msg="Updated message  -> workflowtemplates.argoproj.io \"workflow-template-submittable\" not found" namespace=stackstorm workflow=workflow-template-hello-world-lf6rp                                                                                                                                                                                                      time="2020-08-13T05:47:35Z" level=info msg="Marking workflow completed" namespace=stackstorm workflow=workflow-template-hello-world-lf6rp
time="2020-08-13T05:47:35Z" level=info msg="Checking daemoned children of " namespace=stackstorm workflow=workflow-template-hello-world-lf6rp
I0813 05:47:35.643664       1 request.go:1017] Request Body: {"kind":"Workflow","apiVersion":"argoproj.io/v1alpha1","metadata":{"name":"workflow-template-hello-world-lf6rp","generateName":"workflow-tem
plate-hello-world-","namespace":"stackstorm","selfLink":"/apis/argoproj.io/v1alpha1/namespaces/stackstorm/workflows/workflow-template-hello-world-lf6rp","uid":"4ea1e0d6-1df3-4289-9f76-8d761226a45d","re
sourceVersion":"14244901","generation":1,"creationTimestamp":"2020-08-13T05:47:35Z","labels":{"workflows.argoproj.io/completed":"true","workflows.argoproj.io/phase":"Error"}},"spec":{"arguments":{},"wo
rkflowTemplateRef":{"name":"workflow-template-submittable"}},"status":{"phase":"Error","startedAt":"2020-08-13T05:47:35Z","finishedAt":"2020-08-13T05:47:35Z","message":"workflowtemplates.argoproj.io \"
workflow-template-submittable\" not found","storedTemplates":{"namespaced/workflow-template-submittable/whalesay-template":{"name":"whalesay-template","arguments":{},"inputs":{"parameters":[{"name":"me
ssage"}]},"outputs":{},"metadata":{},"container":{"name":"","image":"docker/whalesay","command":["cowsay"],"args":["{{inputs.parameters.message}}"],"resources":{}}}},"conditions":[{"type":"Completed","
status":"True"}]}}
I0813 05:47:35.643826       1 round_trippers.go:423] curl -k -v -XPUT  -H "Accept: application/json, */*" -H "Content-Type: application/json" -H "Authorization: Bearer eyJhbGciOiJSUzI1NiIsImtpZCI6Im12T
zFnZDhyNTJvSWg4WjVZV0wtZzJmNkgzbTVId2Z1WHF2TEdWUUdMcncifQ.eyJpc3MiOiJrdWJlcm5ldGVzL3NlcnZpY2VhY2NvdW50Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9uYW1lc3BhY2UiOiJhcmdvIiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWN
jb3VudC9zZWNyZXQubmFtZSI6ImFyZ28tdG9rZW4td2dndG0iLCJrdWJlcm5ldGVzLmlvL3NlcnZpY2VhY2NvdW50L3NlcnZpY2UtYWNjb3VudC5uYW1lIjoiYXJnbyIsImt1YmVybmV0ZXMuaW8vc2VydmljZWFjY291bnQvc2VydmljZS1hY2NvdW50LnVpZCI6IjZh
YWJjODM5LTJhYTQtNDk5Ni1iM2MyLWM4YTgyNTMyNThlMSIsInN1YiI6InN5c3RlbTpzZXJ2aWNlYWNjb3VudDphcmdvOmFyZ28ifQ.qECCvvUjjRAAtS1P7cWgkIhcUhw5cgbZrzAdE5aoSC3Ce3OsYYcQTGB2WF7o1r2SDaJXA_vaIbgJog5uy9-8vw_ku9J3qMGENJ
g4Qhz5XC4T_FCeuVj1zLdcT0mltfPw7_V5MRweok9wmjWF7dtk9a3ELFS7aoJ0_HFWXhXeqz2yP3PlOykbfvllD0BF5zS4reihkOFTE2ad3309JDw80GbAMlJQ8eQcX9PSdkyAqrS9OxCoNtYqf-Ctxdi2fGS3HeRxDvGDxi2wwRdPBwbVToJoPaRhBNJmeQb4NDTeJ9Q
iVNwg3J1YfEan7BnhJrx7St2othGs0FPoQEgOMOaujg" -H "User-Agent: workflow-controller/v0.0.0 (linux/amd64) kubernetes/$Format" 'https://172.30.0.1:443/apis/argoproj.io/v1alpha1/namespaces/stackstorm/workflo
ws/workflow-template-hello-world-lf6rp'
I0813 05:47:35.651526       1 round_trippers.go:443] PUT https://172.30.0.1:443/apis/argoproj.io/v1alpha1/namespaces/stackstorm/workflows/workflow-template-hello-world-lf6rp 200 OK in 7 milliseconds
I0813 05:47:35.651539       1 round_trippers.go:449] Response Headers:
I0813 05:47:35.651543       1 round_trippers.go:452]     Audit-Id: 50bd8ca3-12e1-41a8-8d00-6e803614427a
I0813 05:47:35.651546       1 round_trippers.go:452]     Content-Type: application/json
I0813 05:47:35.651549       1 round_trippers.go:452]     Content-Length: 1164
I0813 05:47:35.651551       1 round_trippers.go:452]     Date: Thu, 13 Aug 2020 05:47:35 GMT
I0813 05:47:35.651577       1 request.go:1017] Response Body: {"apiVersion":"argoproj.io/v1alpha1","kind":"Workflow","metadata":{"creationTimestamp":"2020-08-13T05:47:35Z","generateName":"workflow-temp
late-hello-world-","generation":2,"labels":{"workflows.argoproj.io/completed":"true","workflows.argoproj.io/phase":"Error"},"name":"workflow-template-hello-world-lf6rp","namespace":"stackstorm","resour
ceVersion":"14244902","selfLink":"/apis/argoproj.io/v1alpha1/namespaces/stackstorm/workflows/workflow-template-hello-world-lf6rp","uid":"4ea1e0d6-1df3-4289-9f76-8d761226a45d"},"spec":{"arguments":{},"workflowTemplateRef":{"name":"workflow-template-submittable"}},"status":{"conditions":[{"status":"True","type":"Completed"}],"finishedAt":"2020-08-13T05:47:35Z","message":"workflowtemplates.argoproj.io
\"workflow-template-submittable\" not found","phase":"Error","startedAt":"2020-08-13T05:47:35Z","storedTemplates":{"namespaced/workflow-template-submittable/whalesay-template":{"arguments":{},"containe
r":{"args":["{{inputs.parameters.message}}"],"command":["cowsay"],"image":"docker/whalesay","name":"","resources":{}},"inputs":{"parameters":[{"name":"message"}]},"metadata":{},"name":"whalesay-templat
e","outputs":{}}}}}                                                                                                                                                                                      time="2020-08-13T05:47:35Z" level=info msg="Workflow update successful" namespace=stackstorm phase=Error resourceVersion=14244902 workflow=workflow-template-hello-world-lf6rp
time="2020-08-13T05:47:36Z" level=info msg="Released all acquired locks" key=workflow-template-hello-world-lf6rp
```

**Comment by sarabala1979** [comment]
```
Processing workflow" namespace=stackstorm workflow=workflow-template-hello-world-lf6rp
time="2020-08-13T05:47:35Z" level=info msg="Updated message  -> workflowtemplates.argoproj.io \"workflow-template-submittable\" not found" namespace=stackstorm workflow=workflow-template-hello-world-lf6rp             
```
As per log, workflow created in  `namespace=stackstorm` namespace which doesn't have workflowtemplate.

**Comment by dove-young** [comment]
@sarabala1979  sorry for mis-leading.  I mean the namespace `stackstorm`.   I changed all `stackstorm` to `my-namespace` to desensitize information.  But obviously I missed the `stackstorm`in the log. 

I did created workflow template in namespace `stackstorm` and I was refer to the template in workflow in the same namespace.

**Comment by sarabala1979** [comment]
Is it working for you now?

**Comment by dove-young** [comment]
I retried it from scratch, the same error occurs.


- clean up Argo
```
[root@sre4-inf Argo]# oc delete -f install.yaml -n argo
warning: deleting cluster-scoped resources, not scoped to the provided namespace
customresourcedefinition.apiextensions.k8s.io "clusterworkflowtemplates.argoproj.io" deleted
customresourcedefinition.apiextensions.k8s.io "cronworkflows.argoproj.io" deleted
customresourcedefinition.apiextensions.k8s.io "workfloweventbindings.argoproj.io" deleted
customresourcedefinition.apiextensions.k8s.io "workflows.argoproj.io" deleted
customresourcedefinition.apiextensions.k8s.io "workflowtemplates.argoproj.io" deleted
serviceaccount "argo" deleted
serviceaccount "argo-server" deleted
role.rbac.authorization.k8s.io "argo-role" deleted
clusterrole.rbac.authorization.k8s.io "argo-aggregate-to-admin" deleted
clusterrole.rbac.authorization.k8s.io "argo-aggregate-to-edit" deleted
clusterrole.rbac.authorization.k8s.io "argo-aggregate-to-view" deleted
clusterrole.rbac.authorization.k8s.io "argo-cluster-role" deleted
clusterrole.rbac.authorization.k8s.io "argo-server-cluster-role" deleted
rolebinding.rbac.authorization.k8s.io "argo-binding" deleted
clusterrolebinding.rbac.authorization.k8s.io "argo-binding" deleted
clusterrolebinding.rbac.authorization.k8s.io "argo-server-binding" deleted
configmap "workflow-controller-configmap" deleted
service "argo-server" deleted
service "workflow-controller-metrics" deleted
deployment.apps "argo-server" deleted
deployment.apps "workflow-controller" deleted
[root@sre4-inf Argo]#

[root@sre4-inf Argo]# oc get po -n argo
No resources found in argo namespace.
[root@sre4-inf Argo]#
```

- install argo 

```
[root@sre4-inf Argo]# oc apply -f install.yaml -n argo
customresourcedefinition.apiextensions.k8s.io/clusterworkflowtemplates.argoproj.io created
customresourcedefinition.apiextensions.k8s.io/cronworkflows.argoproj.io created
customresourcedefinition.apiextensions.k8s.io/workfloweventbindings.argoproj.io created
customresourcedefinition.apiextensions.k8s.io/workflows.argoproj.io created
customresourcedefinition.apiextensions.k8s.io/workflowtemplates.argoproj.io created
serviceaccount/argo created
serviceaccount/argo-server created
role.rbac.authorization.k8s.io/argo-role created
clusterrole.rbac.authorization.k8s.io/argo-aggregate-to-admin created
clusterrole.rbac.authorization.k8s.io/argo-aggregate-to-edit created
clusterrole.rbac.authorization.k8s.io/argo-aggregate-to-view created
clusterrole.rbac.authorization.k8s.io/argo-cluster-role created
clusterrole.rbac.authorization.k8s.io/argo-server-cluster-role created
rolebinding.rbac.authorization.k8s.io/argo-binding created
clusterrolebinding.rbac.authorization.k8s.io/argo-binding created
clusterrolebinding.rbac.authorization.k8s.io/argo-server-binding created
configmap/workflow-controller-configmap created
service/argo-server created
service/workflow-controller-metrics created
deployment.apps/argo-server created
deployment.apps/workflow-controller created
[root@sre4-inf Argo]#

[root@sre4-inf Argo]# oc get po -n argo
NAME                                  READY   STATUS    RESTARTS   AGE
argo-server-5d8d8d6fff-mtqhq          1/1     Running   0          27s
workflow-controller-84568456f-bjkq7   1/1     Running   0          27s
[root@sre4-inf Argo]#

```


- Create template 

```
[root@sre4-inf Argo]#
[root@sre4-inf Argo]# argo template list -n stackstorm
NAME
[root@sre4-inf Argo]# argo template list -n argo
NAME
[root@sre4-inf Argo]#
[root@sre4-inf Argo]#
[root@sre4-inf Argo]# cd -
/root/Argo/argo/examples/workflow-template
[root@sre4-inf workflow-template]# ls
dag.yaml  hello-world.yaml  retry-with-steps.yaml  steps.yaml  templates.yaml  workflow-template-ref-with-entrypoint-arg-passing.yaml  workflow-template-ref.yaml
[root@sre4-inf workflow-template]# argo template create templates.yaml -n stackstorm
Name:                workflow-template-whalesay-template
Namespace:           stackstorm
Created:             Sun Aug 16 20:05:08 -0700 (1 minute from now)
Name:                workflow-template-random-fail-template
Namespace:           stackstorm
Created:             Sun Aug 16 20:05:08 -0700 (1 minute from now)
Name:                workflow-template-inner-steps
Namespace:           stackstorm
Created:             Sun Aug 16 20:05:08 -0700 (1 minute from now)
Name:                workflow-template-inner-dag
Namespace:           stackstorm
Created:             Sun Aug 16 20:05:08 -0700 (1 minute from now)
Name:                workflow-template-submittable
Namespace:           stackstorm
Created:             Sun Aug 16 20:05:09 -0700 (1 minute from now)
[root@sre4-inf workflow-template]#
```

- Submit workflow 

```
[root@sre4-inf workflow-template]#
[root@sre4-inf workflow-template]# argo submit -n stackstorm workflow-template-ref.yaml
Name:                workflow-template-hello-world-7j425
Namespace:           stackstorm
ServiceAccount:      default
Status:              Pending
Created:             Sun Aug 16 20:05:48 -0700 (1 minute from now)
[root@sre4-inf workflow-template]#
[root@sre4-inf workflow-template]# argo get -n stackstorm workflow-template-hello-world-7j425
Name:                workflow-template-hello-world-7j425
Namespace:           stackstorm
ServiceAccount:      default
Status:              Error
Message:             workflowtemplates.argoproj.io "workflow-template-submittable" not found
Conditions:
 Completed           True
Created:             Sun Aug 16 20:05:48 -0700 (1 minute from now)
Started:             Sun Aug 16 20:05:48 -0700 (1 minute from now)
Finished:            Sun Aug 16 20:05:48 -0700 (1 minute from now)
Duration:            0 seconds
[root@sre4-inf workflow-template]#
```

**Comment by dove-young** [comment]
I tried from a new namespace, the same error occurs

```
[root@sre4-inf workflow-template]# oc new-project test
Now using project "test" on server "https://my-server:6443".

You can add applications to this project with the 'new-app' command. For example, try:

    oc new-app ruby~https://github.com/sclorg/ruby-ex.git

to build a new example application in Python. Or use kubectl to deploy a simple Kubernetes application:

    kubectl create deployment hello-node --image=gcr.io/hello-minikube-zero-install/hello-node

[root@sre4-inf workflow-template]#
[root@sre4-inf workflow-template]#
[root@sre4-inf workflow-template]# argo template create templates.yaml -n test
Name:                workflow-template-whalesay-template
Namespace:           test
Created:             Sun Aug 16 20:07:16 -0700 (1 minute from now)
Name:                workflow-template-random-fail-template
Namespace:           test
Created:             Sun Aug 16 20:07:16 -0700 (1 minute from now)
Name:                workflow-template-inner-steps
Namespace:           test
Created:             Sun Aug 16 20:07:16 -0700 (1 minute from now)
Name:                workflow-template-inner-dag
Namespace:           test
Created:             Sun Aug 16 20:07:17 -0700 (1 minute from now)
Name:                workflow-template-submittable
Namespace:           test
Created:             Sun Aug 16 20:07:17 -0700 (1 minute from now)
[root@sre4-inf workflow-template]#
[root@sre4-inf workflow-template]#
[root@sre4-inf workflow-template]# argo submit -n test  workflow-template-ref.yaml
Name:                workflow-template-hello-world-f2zcg
Namespace:           test
ServiceAccount:      default
Status:              Pending
Created:             Sun Aug 16 20:07:24 -0700 (1 minute from now)
[root@sre4-inf workflow-template]#
[root@sre4-inf workflow-template]# argo get -n test workflow-template-hello-world-f2zcg
Name:                workflow-template-hello-world-f2zcg
Namespace:           test
ServiceAccount:      default
Status:              Error
Message:             workflowtemplates.argoproj.io "workflow-template-submittable" not found
Conditions:
 Completed           True
Created:             Sun Aug 16 20:07:24 -0700 (1 minute from now)
Started:             Sun Aug 16 20:07:24 -0700 (1 minute from now)
Finished:            Sun Aug 16 20:07:24 -0700 (1 minute from now)
Duration:            0 seconds
[root@sre4-inf workflow-template]#
```

**Comment by sarabala1979** [comment]
can you check default service account has RBAC to get and list workflowtemplate kind?

**Comment by sarabala1979** [comment]
Have you joined in Argo Slack? you can find me in sarabal1979(bala). we can chat over there

**Comment by dove-young** [comment]
@sarabala1979 I bound service acconut default to `cluster-admin`
```
oc adm policy add-cluster-role-to-user cluster-admin -n $NS -z default
```

**Comment by sarabala1979** [comment]
Thanks for all details and your help. Found a bug in the master branch(latest). Recent Informer change is causing this issue. I will work on fix.

Please install the stable release and try `kubectl apply -n argo -f https://raw.githubusercontent.com/argoproj/argo/v2.9.5/manifests/install.yaml`

**Comment by alexec** [comment]
@sarabala1979 this fix in #3795 will be for v2.11. I think this bug is for v2.9?
